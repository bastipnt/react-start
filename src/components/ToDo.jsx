// @flow
import React, { Component } from 'react';
import { observer } from 'mobx-react';
import { TLItem, TLEdit, TLShow, Checkbox, Button } from 'styles';

type Props = {
  toDo: TODO,
  deleteToDo: (toDo: TODO) => void,
}

@observer
class ToDo extends Component<Props> {
  handleDoubleClick = (): void => this.props.toDo.edit()

  handleKeyPress = (e: SyntheticEvent<HTMLInputElement>): void => {
    const { toDo } = this.props;
    const target = e.currentTarget;
    if (e.key === 'Enter') {
      toDo.change(target.value);
      target.value = '';
    }
  }

  handleBlur = (): void => {
    const { toDo } = this.props;
    toDo.cancelEdit();
  }

  handleCheckboxChange = (): void => {
    const { toDo } = this.props;
    toDo.toggleCompleted();
  }

  handleDeleteClick = (): void => {
    const { deleteToDo, toDo } = this.props;
    deleteToDo(toDo);
  }

  render = () => {
    const { text, inEdit, completed } = this.props.toDo;

    const editMode = (
      <TLEdit
        defaultValue={text}
        onKeyPress={this.handleKeyPress}
        onBlur={this.handleBlur}
        autoFocus
      />
    );

    const showMode = <TLShow onDoubleClick={this.handleDoubleClick}>{text}</TLShow>;

    return (
      <TLItem completed={completed}>
        <Checkbox
          key="checkbox"
          onChange={this.handleCheckboxChange}
          checked={completed}
        />
        {inEdit ? editMode : showMode}
        <Button key="delete" onClick={this.handleDeleteClick}>
          <i className="fa fa-times" />
        </Button>
      </TLItem>
    );
  }
}

export default ToDo;
