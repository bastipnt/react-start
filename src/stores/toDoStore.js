// @flow
import { observable } from 'mobx';

class ToDo implements TODO {
  @observable text;
  @observable inEdit;
  @observable completed;
  createdAt;
  changedAt;

  constructor(text: string) {
    this.text = text;
    this.createdAt = Date.now();
    this.inEdit = false;
    this.completed = false;
  }

  change = (text: string) => {
    this.text = text;
    this.changedAt = Date.now();
    this.inEdit = false;
    this.completed = false;
  }

  edit = () => { this.inEdit = true; }

  cancelEdit = () => { this.inEdit = false; }

  toggleCompleted = () => { this.completed = !this.completed; }
}

class ToDoStore implements TODOSTORE {
  @observable toDos = []

  createToDo = (text: string) => { if (text) this.toDos.push(new ToDo(text)); }

  deleteToDo = (toDo: TODO) => { this.toDos.remove(toDo); }
}

export default new ToDoStore();
