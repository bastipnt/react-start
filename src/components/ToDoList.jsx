// @flow
import React, { Component } from 'react';
import type { Element } from 'react';
import { inject, observer } from 'mobx-react';
import {
  Page,
  Input,
  TLHeader,
  TLNew,
  TLList,
  TLDate,
  TLMonth,
  TLCount,
  TLWrapper,
} from 'styles';

import ToDo from './ToDo';
import { toDay, toWeekDay, toMonth } from '../utils/dateFormat';

type Props = {
  toDoStore: TODOSTORE,
};

@inject('toDoStore')
@observer
class ToDoList extends Component<Props> {
  handleKeyPress = (e: SyntheticEvent<HTMLInputElement>): void => {
    const { toDoStore } = this.props;
    const target = e.currentTarget;
    if (e.key === 'Enter') {
      toDoStore.createToDo(target.value);
      target.value = '';
    }
  };

  toDos = (): Element<typeof ToDo>[] =>
    this.props.toDoStore.toDos.sort((a) => a.completed ? 1 : -1).map((toDo) =>
      <ToDo key={toDo.createdAt} toDo={toDo} deleteToDo={this.props.toDoStore.deleteToDo} />)

  render = () => {
    const now = new Date();
    return (
      <Page>
        <TLWrapper>
          <TLHeader>
            <TLDate><b>{toWeekDay(now)}</b>, {toDay(now)}th</TLDate>
            <TLMonth>{toMonth(now)}</TLMonth>
            <TLCount><b>{this.props.toDoStore.toDos.length}</b> Tasks</TLCount>
          </TLHeader>
          <TLNew>
            <Input
              id="new-todo"
              onKeyPress={this.handleKeyPress}
              placeholder="Add new task!"
              autoFocus
            />
          </TLNew>
          <TLList>
            {this.toDos()}
          </TLList>
        </TLWrapper>
      </Page>
    );
  }
}

export default ToDoList;
