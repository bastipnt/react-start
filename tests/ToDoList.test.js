import React from 'react';
import { shallow } from 'enzyme';

import {
  TLDate,
  TLMonth,
  Input,
  TLNew,
} from 'styles';
import { mockDate } from '../config/mocks/date.mock';
import ToDoList from '../src/components/ToDoList';
import ToDo from '../src/components/ToDo';
import toDoStore from '../src/stores/toDoStore';

let component;

describe('<ToDoList />', () => {
  describe('content', () => {
    beforeAll(() => {
      mockDate('2017-10-20T00:00:00');
      component = shallow(<ToDoList.wrappedComponent toDoStore={toDoStore} />);
    });

    it('renders correctly', () => {
      expect(component.getElements()).toMatchSnapshot();
    });

    it('should show todays date', () => {
      expect(component.find(TLDate).dive().text()).toBe('Friday, 20th');
      expect(component.find(TLMonth).dive().text()).toBe('October');
    });
  });

  describe('actions', () => {
    beforeEach(() => {
      component = shallow(<ToDoList.wrappedComponent toDoStore={toDoStore} />);
    });

    afterEach(() => {
      toDoStore.toDos = [];
    });

    it('should not add a toDo when the input is empty', () => {
      component.find(TLNew).find(Input).simulate('keyPress', { key: 'Enter', currentTarget: { value: '' } });
      expect(toDoStore.toDos.length).toBe(0);
    });

    it('should not add a toDo when the key pressed is not "Enter"', () => {
      const text = 'Test';
      component.find(TLNew).find(Input).simulate('keyPress', { key: 'A', currentTarget: { value: text } });
      expect(toDoStore.toDos.length).toBe(0);
    });

    it('should add a toDo when the input filled', () => {
      const text = 'Test';
      component.find(TLNew).find(Input).simulate('keyPress', { key: 'Enter', currentTarget: { value: text } });
      expect(toDoStore.toDos.length).toBe(1);
      expect(toDoStore.toDos[0].text).toEqual(text);
      expect(component.find(ToDo).length).toBe(1);
    });

    it('should sort the ToDos in the correct order', () => {
      const text1 = 'Test1';
      const text2 = 'Test2';
      toDoStore.createToDo(text1);
      toDoStore.createToDo(text2);
      component = shallow(<ToDoList.wrappedComponent toDoStore={toDoStore} />);
      expect(component.find(ToDo).length).toBe(2);
      expect(component.find(ToDo).map((toDo) => toDo.render().find('p').text())).toEqual([text1, text2]);
      toDoStore.toDos = [];
      toDoStore.createToDo(text1);
      toDoStore.createToDo(text2);
      toDoStore.toDos[0].completed = true;
      component = shallow(<ToDoList.wrappedComponent toDoStore={toDoStore} />);
      expect(component.find(ToDo).map((toDo) => toDo.render().find('p').text())).toEqual([text2, text1]);
    });
  });
});
