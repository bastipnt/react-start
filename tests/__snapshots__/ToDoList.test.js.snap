// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDoList /> content renders correctly 1`] = `
Array [
  <Layout__Page>
    <ToDoListstyles__TLWrapper>
      <ToDoListstyles__TLHeader>
        <ToDoListstyles__TLDate>
          <b>
            Friday
          </b>
          , 
          20
          th
        </ToDoListstyles__TLDate>
        <ToDoListstyles__TLMonth>
          October
        </ToDoListstyles__TLMonth>
        <ToDoListstyles__TLCount>
          <b>
            0
          </b>
           Tasks
        </ToDoListstyles__TLCount>
      </ToDoListstyles__TLHeader>
      <ToDoListstyles__TLNew>
        <Input
          autoFocus={true}
          id="new-todo"
          onKeyPress={[Function]}
          placeholder="Add new task!"
        />
      </ToDoListstyles__TLNew>
      <List__UL />
    </ToDoListstyles__TLWrapper>
  </Layout__Page>,
]
`;
